r/src/parquet.cpp

```diff
@@ -19,44 +19,101 @@
+    bool use_threads) {
+  return properties->use_threads();
+}
 
+// [[arrow::export]]
+bool parquet___arrow___ArrowReaderProperties__get_read_dictionary(
+    const std::shared_ptr<parquet::arrow::ArrowReaderProperties>& properties,
+    int column_index) {
+  return properties->read_dictionary(column_index);
+}
+
+// [[arrow::export]]
+void parquet___arrow___ArrowReaderProperties__set_read_dictionary(
+    const std::shared_ptr<parquet::arrow::ArrowReaderProperties>& properties,
+    int column_index, bool read_dict) {
+  properties->set_read_dictionary(column_index, read_dict);
+}
+
+// [[arrow::export]]
+std::unique_ptr<parquet::arrow::FileReader> parquet___arrow___FileReader__OpenFile(
+    const std::shared_ptr<arrow::io::RandomAccessFile>& file,
+    const std::shared_ptr<parquet::arrow::ArrowReaderProperties>& props) {
   std::unique_ptr<parquet::arrow::FileReader> reader;
   PARQUET_THROW_NOT_OK(
-      parquet::arrow::OpenFile(infile, arrow::default_memory_pool(), &reader));
+      parquet::arrow::OpenFile(file, arrow::default_memory_pool(), *props, &reader));
+  return reader;
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::Table> parquet___arrow___FileReader__ReadTable1(
+    const std::unique_ptr<parquet::arrow::FileReader>& reader) {
   std::shared_ptr<arrow::Table> table;
   PARQUET_THROW_NOT_OK(reader->ReadTable(&table));
-
   return table;
-#else
-  Rcpp::stop("Support for Parquet is not available.");
+}
 
+// [[arrow::export]]
+std::shared_ptr<arrow::Table> parquet___arrow___FileReader__ReadTable2(
+    const std::unique_ptr<parquet::arrow::FileReader>& reader,
+    const std::vector<int>& column_indices) {
   std::shared_ptr<arrow::Table> table;
+  PARQUET_THROW_NOT_OK(reader->ReadTable(column_indices, &table));
   return table;
-#endif
 }
 
 // [[arrow::export]]
 void write_parquet_file(const std::shared_ptr<arrow::Table>& table,
                         std::string filename) {
-#ifdef ARROW_R_WITH_PARQUET
   std::shared_ptr<arrow::io::OutputStream> sink;
   PARQUET_THROW_NOT_OK(arrow::io::FileOutputStream::Open(filename, &sink));
   PARQUET_THROW_NOT_OK(parquet::arrow::WriteTable(*table, arrow::default_memory_pool(),
                                                   sink, table->num_rows()));
-#else
-  Rcpp::stop("Support for Parquet is not available.");
-#endif
 }
+
+// [[arrow::export]]
+std::shared_ptr<arrow::Schema> parquet___arrow___FileReader__GetSchema2(
+    const std::unique_ptr<parquet::arrow::FileReader>& reader,
+    const std::vector<int>& indices) {
+  std::shared_ptr<arrow::Schema> schema;
+  STOP_IF_NOT_OK(reader->GetSchema(indices, &schema));
+  return schema;
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::Schema> parquet___arrow___FileReader__GetSchema1(
+    const std::unique_ptr<parquet::arrow::FileReader>& reader) {
+  // FileReader does not have this exposed
+  // std::shared_ptr<arrow::Schema> schema;
+  // STOP_IF_NOT_OK(reader->GetSchema(&schema));
```

Comment:
is it worth opening an issue about this, i.e. adding a `::arrow::Status GetSchema(std::shared_ptr<::arrow::Schema>* out);` method to `parquet::arrow::FileReader` ?

Reply:
https://github.com/apache/arrow/pull/4668

Reply:
This PR isn't blocked by that, right? Looks like you could clean this up if that merges first but it's not contingent.

Reply:
It's not, I've used a workaround for now.